recursion/reading-recursive-code.md

```diff
@@ -0,0 +1,288 @@
+| System stack                 | A part of memory dedicated to storing the method calls and local variables of a running program. The system stack is responsible for keeping track of function calls. |
+
+## Recursion
+
+In programming, a recursive function is a function that calls itself.
+
+We can imagine a recursive function named `get_next_item`. When we call `get_next_item`, it will invoke itself, `get_next_item`. Of course, when _this_ function calls begins execution, it will also invoke itself.
+
+Both iteration and recursion are programming techniques used to repeat logic and behavior. There are often iterative and recursive solutions to the same problem!
+
+When would we choose to solve a problem with recursion? Once we're used to thinking about recursion, recursive algorithms may be more readable and understandable compared to iterative solutions.
+
+## Anatomy of a Recursive Function
+
+Functions that call itself, like the imaginary `get_next_item` function, will recurse an infinite number of times.
+
+However, if a function calls itself an infinite number of times, when will we actually get an answer and solve our problem? Recursive functions are more helpful when they actually _do_ stop!
+
+In order to better read and debug recursive functions, let's dissect the anatomy of one.
+
+### Base Cases are Stopping Conditions
+
+Recursive algorithms will usually take in at least one argument. A recursive algorithm handles different situations based on the value of the argument. The algorithm should handle:
+
+- At least one _recursive case_
+- At least one _base case_
+
+| Part           | Definition                                                            |
+| -------------- | --------------------------------------------------------------------- |
+| Recursive Case | A case or situation that depends on the solution of other cases       |
+| Base Case      | A case or situation where there is a solution without using recursion |
+
+The responsibility of the _recursive case_ is to divide the problem into a smaller problem.
+
+The responsibility of the _base case_ is to give a solid, immediate solution to the problem.
+
+We can imagine that a recursive function will handle recursive cases by calling itself, until it reaches a base case. Most recursive functions will follow this pattern:
+
+1. If some condition is met and we're handling the base case
+   - Solve using the base case's solution
+1. Else
+   - Solve by recursively calling the same function with different input
+
+The "different input" in the recursive call should be working towards bringing the problem to the base case.
+
+## Example: Exponentiation
+
+Read through this recursive function, `pow`, whose responsibility is to calculate [exponentiation](https://en.wikipedia.org/wiki/Exponentiation), otherwise known as `num`<sup>`exponent`</sup>. It takes in two arguments:
```

Comment:
prefer power to pow (there is a standard function called pow)
show the params so that num and exponent have context
```suggestion
Read through this recursive function, `power(num, exponent)`, whose responsibility is to calculate [exponentiation](https://en.wikipedia.org/wiki/Exponentiation) of the form `num`<sup>`exponent`</sup>. It takes in two arguments:
```